### DISCUSSION TITLE
Is buffer behavior changed in v1.8?

### DISCUSSION THREAD

**Discussion by yikhim** [shapely/shapely#1405] [Q&A] [answered]
Is the buffer from different segments of a linestring object no longer union with others on version 1.8.*?
Like the example here [https://shapely.readthedocs.io/en/stable/manual.html#object.buffer](https://shapely.readthedocs.io/en/stable/manual.html#object.buffer). I expect the buffer from different segments union with each other and form a polygon.

Here is my example snippet. In this example, as the line geometry is too small compared with the buffer size required, I do not expect an inner ring (hole?) in the output polygon.

Edit: I try to keep my example minimal by drawing a simpler clockwise closed linestring. Also updating the package from 1.8.0 to 1.8.2
Edit2: Specifying cap and join style to make the example mathematically nicer
Edit3: Specifying package installation information required in submission of issues

```
import shapely.geometry
import math

foo = shapely.geometry.LineString([(-20, 0), (0, 20), (20, 0), (0, -20), (-20, 0)])
bar = foo.buffer(100/math.sqrt(2), cap_style=3, join_style=2)

print(shapely.__version__)
print()
print(foo)
print()
print(bar)
print()
print(len(bar.interiors))
```

From an older version, I got what I expected, an inflated diamond, despite the small numerical inaccuracy of an extra point on (-70, -49.99999999999999), which is not much of a problem as that point is basically on the line segment of [(0, -120), (-120, 0)].

The package was from conda with
python                    3.8.13               h6244533_0
geos                      3.8.0                h33f27b4_0
shapely                   1.7.1            py38h06580b3_0

Result running on Windows 10 Home 64 bit
```
1.7.1

LINESTRING (-20 0, 0 20, 20 0, 0 -20, -20 0)

POLYGON ((-120 0, 0 120, 120 0, 0 -120, -70 -49.99999999999999, -120 0))

0
```

But it now contains an inner ring (hole?)

The package was from conda with
python                    3.10.4          h9a8a25e_0_cpython    conda-forge
geos                      3.10.2               h9c3ff4c_0    conda-forge
shapely                   1.8.2           py310hb974679_1    conda-forge

Result running on CentOS 7 x86-6
```
1.8.2

LINESTRING (-20 0, 0 20, 20 0, 0 -20, -20 0)

POLYGON ((-120 0, 0 120, 120 0, 0 -120, -120 0), (-4.9999999999999964 -4.9999999999999964, 4.9999999999999964 -4.9999999999999964, 4.9999999999999964 4.9999999999999964, -4.9999999999999964 4.9999999999999964, -4.9999999999999964 -4.9999999999999964))

1
```

**Comment by mwtoews** [comment, marked as answer]
The difference from each shapely version depends on the version of the [GEOS library](https://github.com/libgeos/geos/) backend, which is responsible for the actual buffer algorithm (in C++). In general, yes, the buffer algorithm should "buffer" individual line segments and union them together, so the second result does not make sense. JTS also produces the second result, as it shares similar algorithms (ping @dr-jts).

The interesting thing with this example is that it is using a closed linestring. If the last coordinate were changed so that the geometry were _not_ closed, the result no longer has an unexpected interior:
```python
from shapely import wkt

foo = shapely.geometry.LineString([(-20, 0), (0, 20), (20, 0), (0, -20), (-20 - 1e-14, 0)])
bar = foo.buffer(100/math.sqrt(2), cap_style=3, join_style=2)
len(bar.interiors)  # 0
wkt.dumps(bar.simplify(1e-12), rounding_precision=12, trim=True)
# POLYGON ((-120 0, 0 120, 120 0, 0 -120, -120 0))
```

**Reply by yikhim** [reply]
Thanks. That looks like I can avoid it by using another version of GEOS as dependency..

**Reply by mwtoews** [reply]
If necessary, pin or use `mamba` / `conda install "geos=3.8"` until this is resolved upstream.

**Comment by dr-jts** [comment]
I think there's a couple of things going on here.  There was a fix to how mitre joins are computed (https://github.com/locationtech/jts/pull/818), and a fix to avoid some buffer artifacts (https://github.com/libgeos/geos/commit/733bddf9483ad562c86882361d67b2c662af304f).

I notice that the `mitre_limit` is not being set explicitly.  This should default to a value of 5, which then produces the diamond shape expected (but with a hole).  Is it possible that Shapely is setting the `mitre_limit` to 0 by default?  As far as I can tell GEOS still uses 5 as the default `mitre_limit`.

The hole artifact is a result of the "fix" to remove artifacts (ironically).  That's going to require some investigation to see if there's a fix.

**Reply by jorisvandenbossche** [reply]
> Is it possible that Shapely is setting the mitre_limit to 0 by default? As far as I can tell GEOS still uses 5 as the default mitre_limit.

Shapely should be using a default of 5.0:

https://github.com/shapely/shapely/blob/15e748b658f9ff9cf51bfb3fd958ded866478ec4/shapely/geometry/base.py#L543

Actually setting `mitre_limit` to 0.0 for a "mitre" join style triggers an exception from GEOS (`IllegalStateException: Cannot compute offset from zero-length line segment`).

**Reply by dr-jts** [reply]
Odd.  GEOS definitely uses 5 as the `mitre_limit` default.  So not sure what's going on.

**Reply by mwtoews** [reply]
The mitre joins are a bit of red herring, I can confirm this as a regression with both JTS and GEOS; see https://github.com/locationtech/jts/issues/876 for details with a simpler example.

**Comment by dr-jts** [comment]
This is now fixed in GEOS back to version 3.8.
